simpleviewinc/keg-herkin
# Updates and fixes

## Context
* There are a few items that need fixed since the recent updates made to keg-hub repos
* I also added a few cleanup items as needed


## Updates
* `package.json`
  * Updated all dependencies
* `repos/tap/assets/icons/*`
  * Added a few Icons used in the UI
* Updated to use the `useIconProps` hook to fix issues with icons
  * `repos/tap/components/buttons/runTestsButton.js`
  * `repos/tap/components/buttons/saveFileButton.js`
  * `repos/tap/components/definition/definitionsEditor.js`
  * `repos/tap/components/toast/toast.js`
* `repos/tap/components/definition/definitionList.js`
  * Moved a lot of the code into it's own hook file
  * This makes the component more manageable
* `repos/tap/components/definition/definitionListItem.js`
  * Fixed class name
* `repos/tap/components/domStyles/domStyles.js`
  * Refactored this to better follow common react patterns
* `repos/tap/components/iframe/iframe.js` 
  * Did a lot of work to fix the height of the Iframe component
  * In some cases the bottom of the content in the Iframe would get cut off
  * This sets a default height similar to how it was set before
  * But it also listen to a post message from the iframe and gets the height from it
    * As of now this does nothing
    * At some point, we'll need to update the backend api to inject a script into the sent HTML
    * The script to be injected is left in the comments of this file
* `repos/tap/components/list/listHeader.js`
  * Cleaned up the styles a lot, and removed unused code
  * Split out the header Icon to it's own component
* `repos/tap/components/list/listHeaderIcon.js`
  * Code moved from the `listHeader.js` component
* `repos/tap/components/list/simpleList.js`
  * Fixed styles and removed unused code
* `repos/tap/components/sidebar/sidebar.js`
  * Updated to use the `setNativeDriver` helper
* `repos/tap/components/tabbar/tab.js`
  * updated to use `useIconProps` hook
  * Updated to allow passing an Icon component
* `repos/tap/containers/screens/screen.js`
  * Add default Icon to screen tabs
* `repos/tap/hooks/useDefinitionGroups.js`
  * Hook code extracted from `repos/tap/components/definition/definitionList.js`
* `repos/tap/hooks/useDomStyles.js`
  * Refactored this to better follow common react patterns
  * Updated to use `convertToCss` helper from re-theme instead of `useStyleTag`
    * This way we're not calling a hook in a loop, which is considered bad practice
    * Also should not be an issue, because it should only ever be called once
* `repos/tap/hooks/useIconProps.js` 
  * Helper hook to generate the correct props for an Icon
  * Also ensure specific props exist
* `repos/tap/hooks/useStyles.js`
  * No longer used, so it was removed
* `repos/tap/hooks/useToggleAnimate.js`
  * Add call to `setNativeDriver` helper
* `repos/tap/theme/*`
  * Fixed styles as needed due to keg-hub repo updates
* `repos/tap/utils/helpers/setNativeDriver.js`
  * Helper to automatically add the `useNativeDriver` property based on current platform
  * Might be a good to migrate this to keg-components
* `repos/tap/utils/validate/*`
  * Replace method with the same method exported from keg-components

## Tests

### Setup
* Pull branch with `keg herkin && keg pr 50`
* Install the dependencies =>  `yarn install`
* Navigate to a repo with a `herkin.config.js` file and run `keg herkin start`
  * or just use the example =>  `keg herkin start --config ./repos/example/herkin.config.js`
* Navigate to http://herkin-updates-and-fixes.local.keghub.io in your browser

### Test 1
* With the keg-herkin app loaded in the browser, select the `example.feature` from the model, and it should load the default editor screen
  * [Looks like this](https://i.imgur.com/eLj90tt.png)
  * Ensure the icons on the buttons display properly
  * Ensure the step definitions in the definitions list display properly
  * Ensure the Definition sections can toggle open and closed for each section `All, Given, When, Then` 
  * For any step definitions that have an arrow toggle, ensure they toggle as expected
    * [Looks like this](https://i.imgur.com/PwfKII2.png)

### Test 2
* Click the Run tests button at the bottom right
  * The UI shoudl switch to the `Test Results` tab, and you should see the terminal output displayed on the screen
    * Ensure the terminal output is display with the correct height
    * [Looks like this](https://i.imgur.com/Kd7Z1W7.png)

### Test 3
* Once the tests finish running
  * Ensure you can scroll all the way to the bottom of the Iframe, below the terminal output
  * In this case there may be a lot of extra white space at the bottom
    * Not really a big deal right now, and will be fixed in the future
* Reload the page, and it should load again, but the terminal output will not be displayed
  * Only the most recent test results should display
  * Ensure you can scroll all the way to the bottom of the Iframe

### Test 4
* On your host machine, open the path at to the test results file that is being loaded in the UI
  * [Looks like this](https://i.imgur.com/MrYgCb2.png)
* At the bottom of the file add this code
```js
  <script>
    console.log(`Sending height to parent window: ${document.body.scrollHeight}`)
    window.parent !== window &&
      window.parent.postMessage({ herkinIframeHeight: document.body.scrollHeight }, '*')
  </script>
```
* [Looks like this](https://i.imgur.com/E0zoVS1.png)
* Go back to the browser, and reload the Test Results page again
* You should see the console.log, with the height sent over
* You should also see that the Iframe height is set to that height
  * [Looks like this](https://i.imgur.com/WiHbnuP.png)
